Migrate to using @actions/core and @actions/github

wardellbagby opened the pull request:
Four commits here. At a high-level, we accomplish a few things with this PR:

1. Start using @actions/core to retrieve Workflow inputs and log messages out.
2. Start using @actions/github to access the GitHub API.
3. Deprecate providing the "GITHUB_TOKEN" as an env variable and instead have it be supplied as a Workflow input.
4. Update the tests to support these changes.
5. Adds a _ton_ more documentation.
6. Add in a new bundler to combine all of the JS code and its dependencies into a single file.

So we all know my plan for the future, once we've got this PR in a good spot and merged, I plan on following it up with a PR that adds ESLint and Prettier to the project (plus a lint check that runs in CI so that all new code will have to pass the lint and be properly formatted).  That PR will also include the README updates I mentioned before, that'll explain how to contribute to this project and run the tests.

I also want to make a PR that makes it so that every new commit will be forced to run the `npm run package` script, so that every new commit will always include a new bundle. That'll stop us from ever altering the code without also including a new bundle and make sure that they're always in sync. Alternatively, if we don't want that, I can add a Lint check that simply makes sure that every commit has an updated bundle in it. Or I can do both to be super safe. Would love some thoughts on that.

As an extra, if you all would allow me to, I'll also convert this Action to be in Typescript instead of Javascript to get the benefit of strict typing. However, I acknowledge that Typescript isn't everyone's cup of tea so I'd like some confirmation that that's okay before I do it.

Similar to my last PR, when checking out this branch, you'll have to run an `npm install` at least once before running `npm run test` in order to run the tests. 

However, there's at least one thing that needs to be discussed. Since this action now uses external dependencies in its normal usage (`@actions/github` and `@actions/core`), it needs to provide a single JS file to GitHub that includes those dependencies as well as the Action code. That's commonly referred to as "bundling" . This PR can't merge without some sort of bundling step. I added in the `@vercel/ncc` bundler that GitHub recommends for Actions. [You can find more information on bundling JS code for Actions here.](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github)

- [x] Add in a bundler step to before merging

Fixes #22

wardellbagby committed:
Migrate to using @actions/core and @actions/github

- Starts using @actions/core which provides functions for easily parsing
arguments that are provided to this argument and logging.
- Starts using @actions/github which provides a typed Javascript
compatible way of accessing GitHub's APIs.
- Separates out the logic of getting the inputs needed to run this
action from the logic of actually running this action to make it easier
to provide different inputs for testing.
- Adds documentation for the various functions in this action.
- Changes the GitHub token to be provided as an input directly instead
of an environment variable to be more inline with how GitHub Actions are
generally structured.

wardellbagby committed:
Support fetching the GitHub token as an argument and as an env variable

- Historically, this Action has only supported getting the GitHub token
from the environment. This isn't very inline with how most other Actions
operate, but in order to support ease-of-upgrading, this commit
continues to support fetching the token from the environment, while
posting a deprecation warning to the logs.

wardellbagby committed:
Update tests to work with Octokit

- Updates the existing tests to work with the new Octokit implementation
via mocking Octokit directly.
- Adds tests to test the input validation logic.
- Fixes some errors that the tests uncovered around input validation
changes and how GitHub will throw when attempting to retrieve an
optional boolean input.

wardellbagby committed:
Add bundler support

- Now that this project includes dependencies, this adds in the @vercel
NCC bundler to bundle this Action's code and all of its dependencies
into a single file that can be ran as the GitHub Action. Also updates
the action.yml file to point to this new bundle.
- Updates the README with instructions on how to check out this project,
runs the associated tests, generate a new bundle, and publish a new
release.

thadguidry commented on README.md:
````diff
@@ -11,3 +11,47 @@ Add following step to your workflow:
+npm run package
+```
+
+and commiting the resulting changes that will be in the `dist/` folder.
````
```suggestion
and committing the resulting changes that will be in the `dist/` folder.
```

thadguidry commented on README.md:
````diff
@@ -11,3 +11,47 @@ Add following step to your workflow:
+npm test
+```
+
+Since this is a repo for a GitHub Action, **every new commit that changes code must include a new bundle.** You can 
````
```suggestion
Since this is a repo for a [GitHub Action](https://docs.github.com/en/actions), **every new commit that changes code must include a new bundle.** You can 
```

thadguidry commented on README.md:
```diff
@@ -11,3 +11,47 @@ Add following step to your workflow:
+and commiting the resulting changes that will be in the `dist/` folder.
+
+## Releasing
+As this repo is a GitHub Action, releases are done via creating a new tag and pushing that tag to the repo. A new
```
```suggestion
Releases are done by creating a new tag and pushing that tag to the repo. A new
```
Let's try to use Simple English, to make this easily translatable for those that don't know English natively.

thadguidry commented on README.md:
````diff
@@ -11,3 +11,47 @@ Add following step to your workflow:
+```
+
+This will update the version in the `package.json`, create a new commit based off of the master branch, and create a 
+new Git tag named the new version that points to that new commit. Remember the `--follow-tags` flag when running 
````
```suggestion
new Git tag with that version that points to the new commit. Remember to use the `--follow-tags` flag when running 
```

thadguidry reviewed (changes requested):
OK, let's try merging this after making the following changes.

wardellbagby committed:
Update README to increase clarity

- Switches some sentences around to use more easily accessible English.
- Adds more links to information around more technical terms.

Co-authored-by: Thad Guidry <thadmguidry@outlook.com>
